packages/js/product-editor/src/components/header/hooks/use-preview/use-preview.tsx

```diff
@@ -0,0 +1,124 @@
+/**
+ * External dependencies
+ */
+import { Product } from '@woocommerce/data';
+import { Button } from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { MouseEvent } from 'react';
+
+export function usePreview( {
+	productId,
+	disabled,
+	onClick,
+	onSaveSuccess,
+	onSaveError,
+	...props
+}: Omit< Button.AnchorProps, 'aria-disabled' | 'variant' | 'href' > & {
+	productId: number;
+	onSaveSuccess?( product: Product ): void;
+	onSaveError?( error: Error ): void;
+} ): Button.AnchorProps {
+	const anchorRef = useRef< HTMLAnchorElement >();
+
+	const { permalink, productStatus, hasEdits } = useSelect(
+		( select ) => {
+			const { getEditedEntityRecord, hasEditsForEntityRecord } =
+				select( 'core' );
+
+			const product = getEditedEntityRecord< Product | undefined >(
+				'postType',
+				'product',
+				productId
+			);
+
+			return {
+				permalink: product?.permalink,
+				productStatus: product?.status,
+				hasEdits: hasEditsForEntityRecord< boolean >(
+					'postType',
+					'product',
+					productId
+				),
+			};
+		},
+		[ productId ]
+	);
+
+	const { editEntityRecord, saveEditedEntityRecord } = useDispatch( 'core' );
+
+	let previewLink: string | undefined;
+	if ( typeof permalink === 'string' ) {
+		if ( permalink.includes( '?' ) ) {
+			previewLink = `${ permalink }&preview=true`;
+		} else {
+			previewLink = `${ permalink }?preview=true`;
+		}
+	}
+
+	/**
+	 * Overrides the default anchor behaviour when the product has unsaved changes.
+	 * Before redirecting to the preview page all changes are saved and then the
+	 * redirection is performed.
+	 *
+	 * @param event
+	 */
+	async function handleClick( event: MouseEvent< HTMLAnchorElement > ) {
+		if ( typeof onClick === 'function' ) onClick( event );
```

Comment:
Well first of all I had a similar discussion with @mattsherman about this and since that day I had been doing the same thing. I know this is not the TypeScript way of doing this, but it's the safer way over all. For me makes all sense because it does not matter what type of value is been passed that code will never fail, it just never be called.

> If we type check onClick, it may fail silently if someone is passing a non function and these types of errors may be harder to surface prior to shipping.

It'll never fail:
- First: the TypeScript type check doesn't let us to pass anything else than the expected type.
- Second: If for some reason, the value of the expected prop type is changed dynamically for something else, then the code will never execute the passed value as a function. Also if we subscribe a non function value to the `onClick` that means that at the end our feature will never work as expected right, so we are violating the first testing step during the development process.

I'm just curious what did we do when TypeScript did not exist long ago to validate the exact same thing.

But, anyway. I can move back to the falsy value check, but please, we need to make sure we are all onboard with the same approach anywhere.

Reply:
This has less to do with TS and more to do with consumers passing the wrong arguments.  If a consumer passes a bad value:

```
onClick={ someFunc() } // This is errantly passing the return value of this function and not the function itself
```

It fails silently and does not give any errors to the consumer.  I'd prefer that errors are surfaced in this case and not harder to discover.

And since callback functions are typically not something that are easily passed dynamically from server to client, most of these issues would be more easily caught by the consumer during development.

Reply:
Well by the moment we write `onClick={ someFunc() }` if the `someFunc()` returns anything different to the expected type then we should see the first signal there is something wrong at typing time.

In case we are not using TypeScript the signal will happen when firing the click event and the implemented feature will not work. Unit tests, e2e tests, manual tests and dev tests will catch that something wrong is happening.
I always prefer the app to be as much resilient as possible rather than not being resilient enough. Anyone should know `onClick={  }` always expect a callable function and not anything else, same thing with `isDisabled` that should always receive a boolean. 

If we really want to not fails silently I would prefer instead of removing that type check validation use it to throw when and unexpected value is passed, so we can really advice the customer asap and detail what was the real cause.

```typescript
if (onClick) {
  if (typeof onClick !== 'function') throw new Error('Please use a real callable function');
  ...
}
```

but again, I suppose that we don't need a validation like that because of TypeScript right. So is this type definition not enough? Why are we using TypeScript then?

```typescript
type Props = {
  onClick?(event: MouseEvent): void;
}
```

Reply:
>Well by the moment we write onClick={ someFunc() } if the someFunc() returns anything different to the expected type then we should see the first signal there is something wrong at typing time.

Again, this may be outside the realm of TS.  I think might be under the assumption that we will only ever use this internally.  We need to assume consumers outside of WooCommerce core may not be using TS, in which case this would fail silently if they passed anything except a function.

For our own protection against passing bad values, TS works well and we should continue using it.  But ignoring bad values passed in makes debugging more challenging for consumers.

>If we really want to not fails silently I would prefer instead of removing that type check validation use it to throw when and unexpected value is passed, so we can really advice the customer asap and detail what was the real cause.

This is fine and I'd prefer this to no error at all.  Personally I find the browser API errors to be more informative and easier to diagnose, but what you posted is fine provided we also make the string translatable.  This could add a hurdle for non-English speaking developers to debug.